ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,271 @@
+from graphql.language.source import Source
+
+from ddtrace import config
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.constants import ERROR_MSG
+from ddtrace.constants import ERROR_STACK
+from ddtrace.constants import ERROR_TYPE
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.internal.compat import stringify
+from ddtrace.internal.utils import ArgumentError
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils import set_argument_value
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.internal.utils.version import parse_version
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+from ddtrace.pin import Pin
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+if TYPE_CHECKING:
+    from typing import Callable
+    from typing import Dict
+    from typing import Iterable
+    from typing import List
+    from typing import Tuple
+    from typing import Union
+
+    from ddtrace import Span
+
+
+_graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(_graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+config._add("graphql", dict(_default_service="graphql"))
+
+
+def _graphql_resolvers_enabled():
+    """
+    Patching graphql resolvers may produce large traces.
+    DD_TRACE_GRAPHQL_RESOLVERS_ENABLED is disabled by default.
+    """
+    return asbool(os.getenv("DD_TRACE_GRAPHQL_RESOLVERS_ENABLED", default=False))
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(wrap, module_str, func_name, wrapper)
+
+
+def unpatch():
+    if not getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(unwrap, module_str, func_name, wrapper)
+
+    setattr(graphql, "_datadog_patch", False)
+
+
+def _get_patching_candidates():
+    if graphql_version < (3, 0):
+        return [
+            ("graphql.graphql", "execute_graphql", _traced_query),
+            ("graphql.language.parser", "parse", _traced_parse),
+            ("graphql.validation.validation", "validate", _traced_validate),
+            ("graphql.execution.executor", "execute", _traced_execute),
+        ]
+    return [
+        ("graphql.graphql", "graphql_impl", _traced_query),
+        ("graphql.language.parser", "parse", _traced_parse),
+        ("graphql.validation.validate", "validate", _traced_validate),
+        ("graphql.execution.execute", "execute", _traced_execute),
+    ]
+
+
+def _update_patching(operation, module_str, func_name, wrapper):
```

Comment:
TODO: For the future we should refactor this behavior into a helper since it'll consistently be needed any time we use the new wrapping helpers.